src/v/container/include/container/contiguous_range_map.h

```diff
@@ -0,0 +1,300 @@
+ * map.
+ *
+ * The wrapper allows random inserts of elements to the map although this is
+ * not supported by standard vector.
+ *
+ * Underlying container is resized every time its size is not sufficient to
+ * account for emplaced key.
+ *
+ * The contiguous_range_map tolerates gap in the range of keys however number
+ * and size of gaps is proportional to performance penalty hit when incrementing
+ * or decrementing iterators.
+ *
+ * NOTE:
+ * A map can store elements for key's which range starts from value greater than
+ * 0 but it will cause significant memory loss. If we will need to store such a
+ * ranges we may introduce non type template parameter indicating offset.
+ */
+template<std::integral KeyT, typename ValueT>
+class contiguous_range_map {
+public:
+    using value_type = std::pair<const KeyT, ValueT>;
+    using key_type = KeyT;
+    using mapped_type = ValueT;
+
+private:
+    using underlying_t = chunked_vector<std::optional<value_type>>;
+
+private:
+    template<bool Const>
+    class iter {
+    public:
+        using iterator_category = std::bidirectional_iterator_tag;
+        using value_type =
+          typename std::conditional_t<Const, const value_type, value_type>;
+
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+        iter() = default;
+
+        /**
+         * Conversion operator allowing iterator to be converted to
+         * const_iterator, as required by the general iterator contract.
+         */
+        operator iter<true>() const { // NOLINT(hicpp-explicit-conversions)
+            iter<true> ret;
+            ret._container = _container;
+            ret._it = _it;
+            return ret;
+        }
+
+        reference operator*() const { return _it->value(); }
+
+        pointer operator->() const { return &(_it->value()); }
+
+        iter& operator++() {
+            ++_it;
+            skip_to_next_present();
+            return *this;
+        }
+
+        iter& operator--() {
+            --_it;
+            skip_to_prev_present();
+            return *this;
+        }
+
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+
+        iter operator--(int) {
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+
+        bool operator==(const iter& o) const {
+            return std::tie(_it, _container) == std::tie(o._it, o._container);
+        };
+
+        auto operator<=>(const iter& o) const {
+            return std::tie(_it, _container) <=> std::tie(o._it, o._container);
+        };
+
+    private:
+        void skip_to_next_present() {
+            while (_it != _container->_values.end() && !_it->has_value()) {
+                ++_it;
+            }
+        }
+
+        void skip_to_prev_present() {
+            while (_it != _container->_values.begin() && !_it->has_value()) {
```

Comment:
calling the decrement operator on the begin iterator will loop forever since there is a decrement then this loop. Can we either change this to <= or assert in the decrement operator that its not the beginning?